In [2]:
import numpy as np
import math
import matplotlib.pyplot as plt
import pandas as pd
from  numpy.random import binomial as rbinom
from numpy.random import normal as rnorm
from numpy.random import uniform as runi
from numpy.random import gamma as gamma
from scipy.stats import norm,binom,t
import statistics
import itertools


The concert would be held outdoors and the hosting cost is 100, 000. If the concert is successful, the total revenue would be 150, 000. However, in case of severe/rainy weather (which can happen with probability of 20%), the event would have to be cancelled, all the ticket sales reimbursed, and the entire upfront expenses of $100,000 would be lost.  


One unit of insurance costs 30 cents and pays 1 dollar in case severe weather does occur, and nothing otherwise. AS can buy as much or as little insurance as it likes, with the respective cost counted on top of the original $100,000.


Find the expected return for AS from hosting the concert assuming they buy x units of insurance. Note that return is defined here as (Revenue-Cost)/Cost: e.g. if AS buys zero insurance and the weather is good they will earn 150K in return for spending 100K, for a return of 50%. In the same context if the weather is bad, they will have a return of -100%, losing all their money.

In [17]:
def ex_return(x):
    """
    input: x: units of thousands of insurance
    out: expected return
    """
    ex_return = 0.8*((150-100-0.3*x)/(100+0.3*x))+0.2*((x-100-0.3*x)/(100+0.3*x))
    return ex_return


In [18]:
# Find the variance
def var(x):
    """
    input: x: units of thousands of insurance
    out: variance of return
    """
    return_1 = (150-100-0.3*x)/(100+0.3*x)
    return_2 = (x-100-0.3*x)/(100+0.3*x)
    var= 0.8*return_1**2+0.2*return_2**2 - ex_return(x)**2
    return var

In [19]:
#  maximize E[Kv]-1/2Var(Kv)
variance=[]
ex = []
maximizer = []
for k in range(0,100):
    variance.append(var(k))
    ex.append(ex_return(k))
    maximizer.append(ex_return(k)-1/2*var(k))
max_index=maximizer.index(max(maximizer))
x=max_index
print("The percentage of revenue to be insured is", x,"%")
pd.DataFrame({"variance":variance,"expected return":ex,"Maximizer":maximizer})

The percentage of revenue to be insured is 67 %


variance  expected return  Maximizer
0   0.360000         0.200000   0.020000
1   0.353094         0.198405   0.021858
2   0.346296         0.196819   0.023671
3   0.339604         0.195243   0.025441
4   0.333016         0.193676   0.027168
5   0.326531         0.192118   0.028853
6   0.320147         0.190570   0.030496
7   0.313863         0.189030   0.032099
8   0.307678         0.187500   0.033661
9   0.301590         0.185979   0.035183
10  0.295598         0.184466   0.036667
11  0.289700         0.182962   0.038112
12  0.283896         0.181467   0.039519
13  0.278183         0.179981   0.040889
14  0.272560         0.178503   0.042223
15  0.267027         0.177033   0.043520
16  0.261581         0.175573   0.044782
17  0.256223         0.174120   0.046008
18  0.250950         0.172676   0.047201
19  0.245761         0.171239   0.048359
20  0.240655         0.169811   0.049484
21  0.235631         0.168391   0.050576
22  0.230688         0.166979   0.051635
23  0.225825         0.165575   0.052663
24  0.221040         0.164179   0.053659
25  0.216333         0.162791   0.054624
26  0.211702         0.161410   0.055559
27  0.207147         0.160037   0.056463
28  0.202666         0.158672   0.057339
29  0.198258         0.157314   0.058184
..       ...              ...        ...
70  0.069941         0.107438   0.072468
71  0.067866         0.106348   0.072415
72  0.065833         0.105263   0.072347
73  0.063840         0.104184   0.072264
74  0.061888         0.103110   0.072166
75  0.059975         0.102041   0.072053
76  0.058101         0.100977   0.071926
77  0.056266         0.099919   0.071786
78  0.054470         0.098865   0.071631
79  0.052711         0.097817   0.071462
80  0.050989         0.096774   0.071280
81  0.049303         0.095736   0.071084
82  0.047654         0.094703   0.070876
83  0.046041         0.093675   0.070654
84  0.044463         0.092652   0.070420
85  0.042920         0.091633   0.070173
86  0.041411         0.090620   0.069914
87  0.039937         0.089611   0.069643
88  0.038495         0.088608   0.069360
89  0.037087         0.087609   0.069065
90  0.035712         0.086614   0.068758
91  0.034369         0.085625   0.068440
92  0.033058         0.084639   0.068111
93  0.031778         0.083659   0.067770
94  0.030530         0.082683   0.067419
95  0.029312         0.081712   0.067056
96  0.028124         0.080745   0.066683
97  0.026966         0.079783   0.066300
98  0.025838         0.078825   0.065906
99  0.024739         0.077872   0.065503

[100 rows x 3 columns]

# Problem 2

$$

K_i = (S_i(1) - S_i(0) ) /S_i(0)
\\
\alpha_i -r = \beta_i (\alpha_M - r)

\\
E(s_i(1)) =s_i(0)(1+r+\beta_i(\alpha_M-r))

$$

In [20]:
s1_1 = np.array([30,15,25,20,10])
sm_1 = np.array([112,108,104,100,92])
prob = np.array([0.1,0.3,0.4,0.15,0.05])
pd.DataFrame({'S1(1)':s1_1,"SM(1)":sm_1,"probability":prob})


S1(1)  SM(1)  probability
0     30    112         0.10
1     15    108         0.30
2     25    104         0.40
3     20    100         0.15
4     10     92         0.05

CAPM:  $\mu_1 = r+\beta(\mu_M -r)$

$\mu_1 =E[k_1] = \sum k_1(w_i)*p(w_i) = \sum (s_i*p_i /s0) -1$
$\mu_M =E[k_M] = \sum k_m(w_i)*p(w_i) = \sum (s_Mi*p_i /s0) -1$

$\beta=E[k1*kM] - E[K1]E[KM] / Var(Km)$
$E[k1 kM]=\frac{1}{s1(0)} (\sum s1_i(1) K_Mi*p_i) - \sum p_i k_Mi$

In [21]:
r=0.04
#sum of each scenario of s1(1)*probability
sum_sp = np.dot(s1_1,prob)
muM= np.dot(sm_1,prob)/100 -1
# sum of s1(1)*Km*p
a =  sum(prob*(sm_1/100 -1)*s1_1)
# sum of p*kM
b = np.dot(prob,sm_1/100 - 1)
# Variance of KM
var_KM=sum((sm_1/100 -1)**2 * prob) - muM**2
print(a,b,var_KM,sum_sp,muM)

1.080000000000001 0.04800000000000005 0.002016000000000003 21.0 0.04800000000000004


1. From CAPM: $\beta = \frac{\frac{21}{s_1(0)} -1-0.004}{0.048-0.04}$

2. From market risk  $\beta = \frac{1.08}{s_1(0)} -0.048-(\frac{21}{s1(0)}-1)(0.048)/ 0.002016$

## $\Rightarrow$ Thus $s_1(0) = 19.91758$


## Question 3

In [22]:
 data_q3=pd.read_csv("IBM-MSFT-HAS.csv")
data_q3=data_q3.set_index(['Date'])
data_q3a =data_q3.loc["11/1/2017":"4/30/2018"]


In [23]:
# a estimate montly avg return,var, cov
def avg_return(data):
    n = len(data) -1
    k_ibm = np.zeros(n)
    k_msft = np.zeros(n)
    k_has= np.zeros(n)
    ibm=data["IBM"]
    msft=data["MSFT"]
    has =data['HAS']
    for i in range(n):
        k_ibm[i] = (ibm[i+1]/ibm[i]) -1
        k_msft[i] = (msft[i+1]/msft[i]) -1
        k_has[i] = (has[i+1]/has[i]) -1
    avg_return_ibm = 20*np.mean(k_ibm)
    avg_return_msft =20*np.mean(k_msft)
    avg_return_has = 20*np.mean(k_has)
    var_ibm = 20**2*np.var(k_ibm)
    var_msft=20**2*np.var(k_msft)
    var_has = 20**2*np.var(k_has)
    cov = 20**2*np.cov([k_ibm,k_msft,k_has])
    mu = np.array([avg_return_ibm,avg_return_msft,avg_return_has])
    var = np.array([var_ibm,var_msft,var_has])
    return mu,cov,var

In [32]:
# b
# weights of market portfolio, annual interst rate r=0.02
mu,cov,var = avg_return(data_q3a)
r_monthly = (1+0.02)**(1/12)-1

one =np.ones(3)
w_market = np.dot((mu-r_monthly*one),np.linalg.inv(cov))
w=w_market/sum(w_market)
print("Monthly Avg Return\n",mu,"\nVariance\n", var,"\n Covaraince \n",cov)

Monthly Avg Return
 [-0.00454143  0.02369024  0.00076392] 
Variance
 [0.08566291 0.11550066 0.11984941] 
 Covaraince 
 [[0.08637087 0.0568387  0.01617577]
 [0.0568387  0.11645521 0.02203789]
 [0.01617577 0.02203789 0.1208399 ]]


In [25]:
print("Market Porfolio is:", w)

Market Porfolio is: [-15.6413831   18.28327187  -1.64188877]


In [26]:
risk = np.dot(np.dot(w,cov),w.T)
return_m= np.dot(w,mu)
print("Market risk is",risk,"Market return is", return_m)

Market risk is 27.383806494433887 Market return is 0.5029150576174263


In [33]:
# c Repeat first half of data set
data_q3c =data_q3.loc["5/1/2017":"10/31/2017"]
mu,cov,var=avg_return(data_q3c)
w_market = np.dot((mu-r_monthly*one),np.linalg.inv(cov))
w=w_market/sum(w_market)
print("Market Porfolio is:", w)
risk = np.dot(np.dot(w,cov),w.T)
return_m= np.dot(w,mu)
print("Market risk is",risk,"Market return is", return_m)


Market Porfolio is: [-0.04689894  1.4901234  -0.44322446]
Market risk is 0.0887721659395448 Market return is 0.04892844982678925


In [34]:
# d Repeat entire year
mu,cov,var=avg_return(data_q3)
w_market = np.dot((mu-r_monthly*one),np.linalg.inv(cov))
w=w_market/sum(w_market)
print("Market Porfolio is:", w)
risk = np.dot(np.dot(w,cov),w.T)
return_m= np.dot(w,mu)
print("Market risk is",risk,"Market return is", return_m)

Market Porfolio is: [-2.32606204  4.514368   -1.18830596]
Market risk is 1.3599233019839352 Market return is 0.13448840558659883


## 4 
Consider a binomial tree with u = 1.1, d = 0.92, S0 = 20 and r = 0.04. The
physical probability is p = 0.5. We consider an exponential utility maximization problem. Let U(x) = −2e^(−x/2).
Find optial terminal weath X_N that maximize $E[U(X_N)]$ as a function of number of periods N and X0.

$x_i$ : weath at time N in scenario in $i$

There are total $2^N$ scenarios

$x_n^\star =max_{xn} \sum p_i U(x_i) +\lambda(\sum p_i \xi_i xi_i -x_0)$

Take the derivative and equate to 0:
$$ 
x_i^\star = -2*ln(-\lambda^\star \xi_i)
\\
x_0(1+r)^N = \sum q_i(-2 ln(-\lambda^\star \xi))
$$

In [33]:
def expand_grid(data_dict):
    rows = itertools.product(*data_dict.values())
    return pd.DataFrame.from_records(rows,columns=data_dict.keys())

u=1.1
d=0.92
w=expand_grid({"N" + str(x+1):[2/3,1/3] for x in range (3) })
w=w.values 

q_N=w[:,[0]]*w[:,[1]]*w[:,[2]]
q_N

array([[0.2962963 ],
       [0.14814815],
       [0.14814815],
       [0.07407407],
       [0.14814815],
       [0.07407407],
       [0.07407407],
       [0.03703704]])

In [34]:

r=0.04
p=0.5
q = (1+r-d)/(u-d) #2/3
def optimal_weath(N,x0):
    q_list = q_N
    z=np.zeros(8)
    for i in range(8):
        z[i] =math.log(q_list[i]/(p**N*(1+r)**N))
        
    lam =-math.exp((x0*(1+r)**N+2*sum(q_list[i]*z[i]))/(-2*sum(q_list)))
    return lam
# find the optimal lambda
lam=optimal_weath(3,10)
lam

-0.0037918150697505456

In [37]:
N=3
x3 = np.zeros(2**N)
u= np.zeros(2**N)
# Find X3 under 8 different scenarios 
for i in range (2**N):
    x3[i] = -2*math.log(-lam*q_N[i]/(p**N*(1+r)**N))
    
    u[i] = 0.5**N*(-2*math.exp((-x3[i]/2))) #Expected utility
print("Three-period model X3 \n",x3,"\n Expected utility is at N=0",sum(u))


Three-period model X3 
 [ 9.65905277 11.04534713 11.04534713 12.4316415  11.04534713 12.4316415
 12.4316415  13.81793586] 
 Expected utility is at N=0 -0.0067418195795234705
